fabric-resource-loader-v1/src/main/java/net/fabricmc/fabric/impl/resource/v1/ResourceLoaderImpl.java

```diff
@@ -0,0 +1,276 @@
+			.orElse(true);
+
+	public static ResourceLoaderImpl get(ResourceType type) {
+		return IMPL_MAP.computeIfAbsent(type, ResourceLoaderImpl::new);
+	}
+
+	private final Set<Identifier> addedReloaderIds = new ObjectOpenHashSet<>();
+	private final Set<IdentifiableResourceReloader> addedReloaders = new LinkedHashSet<>();
+	private final Set<ReloaderOrder> reloadersOrdering = new LinkedHashSet<>();
+	private final ResourceType type;
+
+	private ResourceLoaderImpl(ResourceType type) {
+		this.type = type;
+	}
+
+	@Override
+	public void registerReloader(IdentifiableResourceReloader reloader) {
+		if (!this.addedReloaderIds.add(reloader.getFabricId())) {
+			throw new IllegalStateException(
+					"Tried to register resource reloader %s twice!".formatted(reloader.getFabricId())
+			);
+		}
+
+		if (!this.addedReloaders.add(reloader)) {
+			throw new IllegalStateException(
+					"Resource reloader with previously unknown ID %s already in resource reloader set!"
+							.formatted(reloader.getFabricId())
+			);
+		}
+	}
+
+	@Override
+	public void addReloaderOrdering(@NotNull Identifier firstReloader, @NotNull Identifier secondReloader) {
+		Objects.requireNonNull(firstReloader, "The first reloader identifier should not be null.");
+		Objects.requireNonNull(secondReloader, "The second reloader identifier should not be null.");
+
+		if (firstReloader.equals(secondReloader)) {
+			throw new IllegalArgumentException("Tried to add a phase that depends on itself.");
+		}
+
+		this.reloadersOrdering.add(new ReloaderOrder(firstReloader, secondReloader));
+	}
+
+	public static List<ResourceReloader> sort(ResourceType type, List<ResourceReloader> listeners) {
+		if (type == null) {
+			return listeners;
+		}
+
+		ResourceLoaderImpl instance = get(type);
+
+		if (instance != null) {
+			var mutable = new ArrayList<>(listeners);
+			instance.sort(mutable);
+			return Collections.unmodifiableList(mutable);
+		}
+
+		return listeners;
+	}
+
+	/**
+	 * Sorts the given resource reloaders to satisfy dependencies.
+	 *
+	 * @param reloaders the resource reloaders to sort
+	 */
+	private void sort(List<ResourceReloader> reloaders) {
+		// Build the actual full list of resource reloaders to add.
+		final Set<IdentifiableResourceReloader> reloadersToAdd = new LinkedHashSet<>(this.addedReloaders);
+
+		// Locate and extract the setup marker.
+		ResourceReloader setupReloader = this.extractSetupMarker(reloaders);
+
+		// Remove any modded reloaders to sort properly.
+		reloaders.removeAll(reloadersToAdd);
+
+		// General rules:
+		// - We *do not* touch the ordering of vanilla reloaders. Ever.
+		//   While dependency values are provided where possible, we cannot
+		//   trust them 100%. Only code doesn't lie.
+		// - We add all custom reloaders after vanilla reloaders if they don't have contrary ordering. Same reasons.
+
+		var runtimePhases = new Object2ObjectOpenHashMap<Identifier, ResourceReloaderPhaseData>();
```

Comment:
Just to make sure, ordering is irrelevant here since it's toposorted and we make no guarantee of order between unordered modded reloaders, right?

Reply:
There is no guarantee of ordering between unordered modded reloaders. The order of this map is irrelevant.